L58-python-async-api.md

```diff
@@ -0,0 +1,453 @@
+Async API for gRPC Python
+----
+* Author(s): lidiz
+* Approver: a11r
+* Status: Draft
+* Implemented in: Python
+* Last updated: 2019-07-22
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/7V7HYM_aph4
+
+## Abstract
+
+A brand new set of async API that will solve concurrency issues and performance
+issues for gRPC Python, which is available to Python 3.6+.
+
+## Motivation
+
+* Asynchronous processing perfectly fits IO-intensive gRPC use cases;
+* Resolve a long-living design flaw of thread exhaustion problem;
+* Performance is much better than the multi-threading model.
+
+## Background
+
+### What Is Asynchronous I/O In Python?
+
+Quote from [`asyncio`](https://docs.python.org/3/library/asyncio.html) package
+documentation:
+
+> `asyncio` is a library to write concurrent code using the `async`/`await`
+> syntax.
+>
+> `asyncio` is used as a foundation for multiple Python asynchronous frameworks
+> that provide high-performance network and web-servers, database connection
+> libraries, distributed task queues, etc.
+
+In the asynchronous I/O model, the computation tasks are packed as generators
+(aka. coroutines) that will yield the ownership of thread while it is blocked by
+IO operations or wait for other tasks to complete.
+
+The design of coroutine in Python can trace back to Python 2.5 in [PEP 342 --
+Coroutines via Enhanced Generators](https://www.python.org/dev/peps/pep-0342/).
+It introduces new generator methods like `send`, `throw`, and `close`. With
+these methods, the caller of the generator can pass values and exceptions into a
+paused generator function, hence diminish the need for wiring complex callbacks.
+
+To further simplify the usage of a coroutine, Python 3.5 introduces
+`async`/`await` syntax that can easily transform a normal function into a
+coroutine ([PEP 492](https://www.python.org/dev/peps/pep-0492/)). And the
+`Twisted` style event loop abstractions is introduced in [PEP
+3156](https://www.python.org/dev/peps/pep-3156/).
+
+Another important piece is the asynchronous generator. It is introduced in
+Python 3.6 by [PEP 525](https://www.python.org/dev/peps/pep-0525/).
+
+To read more about `asyncio` implementation in CPython, see:
+* [C-Level
+  implementation](https://github.com/python/cpython/blob/3.7/Modules/_asynciomodule.c)
+* [Python-Level
+  implementation](https://github.com/python/cpython/blob/3.7/Lib/asyncio)
+
+<!-- ### Benefits Of `asyncio`
+
+* Asynchronous processing that perfectly fits IO-intensive gRPC usage;
+* Resolve a long-living design flaw of thread exhaustion problem;
+* Performance is much better than multi-threaded.
+
+### Drawbacks Of `asyncio`
+
+* Non-interoperability between coroutine functions and normal functions;
+* Cognitive overhead of event loop mechanism;
+* Calls to function may cause context-switches which resulted in unexpected
+  behavior. -->
```

Comment:
WDYM? It's hard to believe that a context switch might have as a side effect a ` unexpected behavior`

Reply:
An example of unexpected behavior will be the need for synchronization. Users of `asyncio` may think that there is only one thread operating, so it is safe to get rid of `Lock`/`Event`/`Condition`.

However, it isn't. Take the following code as example, if the user is trying to operate DB or other global state, data race can still happen.

```Python
import asyncio

class Account:
    def __init__(self, name, amount):
        self.name = name
        self.amount = amount
    def __repr__(self):
        return f"<class Account name={self.name} amount={self.amount}>"

accounts = {
    1: Account(name='John', amount=70),
    2: Account(name='Smith', amount=30),
}

async def withdraw(account_id, diff):
    await asyncio.sleep(0.5)  # Mock async db read operation
    account = accounts[account_id]
    new_amount = account.amount - diff
    if new_amount < 0:
        raise ValueError('Over withdraw!')
    await asyncio.sleep(0.5)  # Mock async db write operation
    account.amount = new_amount
    print(f'Successfully withdraw [{diff}] amount of money')

async def main():
    await asyncio.gather(
        withdraw(2, 30),
        withdraw(1, 45),
        withdraw(1, 30),
        withdraw(2, 30),
    )
    print(accounts)

asyncio.run(main())
```

All withdraw operation will success, which is wrong:
```shell
Successfully withdraw [30] amount of money
Successfully withdraw [45] amount of money
Successfully withdraw [30] amount of money
Successfully withdraw [30] amount of money
{1: <class Account name=John amount=40>, 2: <class Account name=Smith amount=0>}
```

Reply:
Agree, data races can happen and the developers need to take this into account for avoiding them. But Python, and specifically Asyncio provides the primitives as building blocks for implementing these features that would help the developer to address this kind of issues.

Calling out that the chances of having data races is a Asyncio drawbacks it's IMO a bit unfair taking into account that this can happen in other alternatives like gevent, twisted, or even in another languages/frameworks like Node and Go. 

Im wondering if there is any language and framework that providing the capacity of achieving concurrency its at the same time free of having data races, do you have any example?

Reply:
Yes, I agree that data races is inevitable with concurrency. That's why I commented those lines. They shouldn't show with normal view. I will remove them.

Maybe my wording is not accurate here. In my very first draft, I wrote these points as a reminder for myself that introducing `asyncio` into a system might have some side effects that users might not fully aware of. So, during design phase, I can try to protect users from going into these error-prone mistakes.